File: basic-go/leetcode/list/delete_duplicates.go
package list

// 给定排序链表，删除所有重复元素，使得每个元素只出现一次

// 朴素法：当前元素和下一个元素相等，则跳过下一个元素
func deleteDuplicates(head *ListNode) *ListNode {
	curr := head

	for curr != nil && curr.Next != nil {
		if curr.Val == curr.Next.Val {
			tmp := curr.Next
			curr.Next = curr.Next.Next
			tmp.Next = nil
		} else {
			curr = curr.Next
		}
	}

	return head
}

// 使用双指针，记录重复元素开始和结束
// 1. left 和 right 相同则需要同时移动
// 2. left 慢于 right 则中间有相等的元素，需要移除，然后将 left 和 right 相等
func deleteDuplicates1(head *ListNode) *ListNode {

	if head == nil {
		return nil
	}
	left, right := head, head
	for right.Next != nil {
		prev := right
		right = right.Next
		if left.Val != right.Val {
			// 顺序不可调换，prev 有可能和 left 相等
			prev.Next = nil
			left.Next = right
			left = right
		}
	}
	if left.Val == right.Val {
		left.Next = nil
	}
	return head
}

// 给定排序链表，删除所有含有重复数字的节点，只保留原始链表中 没有重复出现 的数字

// 采用递归实现
// 1. 1->1->2->3	需要先将前面的重复去掉
// 2. 1->2->2->3	递归会变成 1

func deleteDuplicates2(head *ListNode) *ListNode {

	// base case
	if head == nil || head.Next == nil {
		return head
	}
	if head.Val == head.Next.Val {
		next := head.Next
		for next != nil && next.Val == head.Val {
			next = next.Next
		}
		return deleteDuplicates(next)
	}
	head.Next = deleteDuplicates(head.Next)
	return head
}

File: basic-algorithm/example/src/leetcode/list/DeleteDuplicates.java
package example.leetcode.list;

/**
 *  给定升序链表头节点 head，删除重复的元素，使得每个元素只出现一次
 */
public class DeleteDuplicates {

    /**
     * 双指针，记录重复出现的开始和结束位置
     */
    public ListNode deleteDuplicates(ListNode head){

        ListNode dummy = new ListNode(Integer.MIN_VALUE, head);
        ListNode left = dummy, right = head;

        while (right != null){
            if (left.val != right.val){
                left.next = right;
                left = right;
            }
            right = right.next;
        }
        if (left.next != null){
            left.next = null;
        }
        return dummy.next;
    }


    /**
     *  给定升序链表头节点 head，只保留没有重复的节点
     *
     *  双指针，记录重复开始的前缀和重复结束的位置
     *
     */
    public ListNode deleteDuplicates_1(ListNode head){
        if (head == null || head.next == null){
            return head;
        }
        ListNode dummy = new ListNode(Integer.MIN_VALUE, head);
        ListNode left = dummy, right = head.next;
        while (right != null){

            if (left.next.val != right.val){
                if (left.next.next == right){
                    left = left.next;
                }else{
                    left.next = right;
                }
            }
            right = right.next;
        }
        if (left.next.next != null){
            left.next = null;
        }
        return dummy.next;
    }
}
